Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift

```diff
@@ -160,6 +160,190 @@ extension UnifiedSymbolGraph {
+            }).sorted(by: Symbol.sortForOverloads(orderByDeclaration: allSymbolsHaveDeclaration))
+            let firstOverload = sortedOverloads.first!
+            let computedOverloadGroup = firstOverload.uniqueIdentifier + SymbolGraph.Symbol.overloadGroupIdentifierSuffix
+
+            guard siblingOverloadGroups.contains(computedOverloadGroup) else {
+                assertionFailure(
+                """
+                UnifiedSymbolGraph.combineOverloadGroups found a candidate overload group that didn't
+                match any existing overload group from its constituent symbol graphs.
+
+                The unified graph found that \(firstOverload.uniqueIdentifier) should be the default
+                overload, but the constituent symbol graphs contained the following overload groups:
+                \(siblingOverloadGroups)
+                """
+                )
+                continue
+            }
+
+            for overloadIndex in sortedOverloads.indices {
+                let overloadedSymbol = sortedOverloads[overloadIndex]
+
+                // 3. Ensure that all the overloaded symbols have `overloadOf` relationships pointing
+                // to the computed overload group.
+                for selector in overloadedSymbol.mainGraphSelectors {
+                    let newRelationship = SymbolGraph.Relationship(
+                        source: overloadedSymbol.uniqueIdentifier,
+                        target: computedOverloadGroup,
+                        kind: .overloadOf,
+                        targetFallback: nil)
+                    if relationshipsByLanguage[selector]?.contains(where: { otherRelationship in
+                        otherRelationship.source == newRelationship.source &&
+                        otherRelationship.target == newRelationship.target &&
+                        otherRelationship.kind == newRelationship.kind
+                    }) != true {
+                        relationshipsByLanguage[selector, default: []].append(newRelationship)
+                    }
+                }
+
+                // 4. Create a new OverloadData instance and save it to the overloaded symbol.
+                let overloadData = SymbolGraph.Symbol.OverloadData(
+                    overloadGroupIdentifier: computedOverloadGroup,
+                    overloadGroupIndex: overloadIndex)
+                overloadedSymbol.unifiedMixins[SymbolGraph.Symbol.OverloadData.mixinKey] = overloadData
+            }
+
+            // 5. If there were any other overload groups that pointed to the same symbols, remove
+            // them from the unified graph, along with any relationships that included them.
+            for siblingOverloadGroup in siblingOverloadGroups {
+                // Remove all the sibling groups from the dictionary we're iterating over so we don't
+                // recalculate all this all over again.
+                overloadGroups.removeValue(forKey: siblingOverloadGroup)
```

Comment:
I had to look this up while writing it so i could be sure. 😅 

The iterator used in a `for ... in` loop is a _copy_ of the collection being iterated over. (More precisely, [Dictionary's iterator is a value type](https://developer.apple.com/documentation/swift/dictionary/iterator) which contains a reference to the original dictionary, which, like the other collections in the Swift standard library, exercises copy-on-write semantics so that the iterator's view of the dictionary doesn't change while the caller's view does.)

By modifying the collection while iterating over it, you don't invalidate the iterator itself. However, you can still create confusing situations, which is why i perform a `guard` at the beginning of the outer loop so that i don't process overload groups that have been removed.

I can see this being cleaned up by creating a new set of `processedOverloadGroups` and checking membership there instead of in the main dictionary, but this is the version i wound up writing. 🤷‍♀️

Reply:
Thanks a lot for the explanation! Super helpful. No, this code is fine as along as we know iterators work this way.